"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.caught = void 0;
var AxePartialRunner = /** @class */ (function () {
    function AxePartialRunner(partialPromise, initiator) {
        if (initiator === void 0) { initiator = false; }
        this.initiator = initiator;
        this.childRunners = [];
        this.partialPromise = (0, exports.caught)(partialPromise);
    }
    AxePartialRunner.prototype.addChildResults = function (childResultRunner) {
        this.childRunners.push(childResultRunner);
    };
    AxePartialRunner.prototype.getPartials = function () {
        return __awaiter(this, void 0, void 0, function () {
            var parentPartial, childPromises, childPartials, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.partialPromise];
                    case 1:
                        parentPartial = _a.sent();
                        childPromises = this.childRunners.map(function (childRunner) {
                            return childRunner ? (0, exports.caught)(childRunner.getPartials()) : [null];
                        });
                        return [4 /*yield*/, Promise.all(childPromises)];
                    case 2:
                        childPartials = (_a.sent()).flat(1);
                        return [2 /*return*/, __spreadArray([parentPartial], childPartials, true)];
                    case 3:
                        e_1 = _a.sent();
                        if (this.initiator) {
                            throw e_1;
                        }
                        return [2 /*return*/, [null]];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return AxePartialRunner;
}());
exports.default = AxePartialRunner;
// Utility to tell NodeJS not to worry about catching promise errors async.
// See: https://stackoverflow.com/questions/40920179/should-i-refrain-from-handling-promise-rejection-asynchronously
exports.caught = (function (f) {
    return function (p) { return (p.catch(f), p); };
    /* eslint-disable @typescript-eslint/no-empty-function */
})(function () { });
//# sourceMappingURL=AxePartialRunner.js.map